File: android/src/main/java/com/gachlab/capacitor/permissions/DevicePermissionsPlugin.java
package com.gachlab.capacitor.permissions;

import android.Manifest;
import android.os.Build;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;

@CapacitorPlugin(
    name = "DevicePermissions",
    permissions = {
        @Permission(
            alias = "geolocation",
            strings = { Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION }
        ),
        @Permission(alias = "notifications", strings = {}),
        @Permission(alias = "notificationsPolicy", strings = { Manifest.permission.ACCESS_NOTIFICATION_POLICY })
    }
)
public class DevicePermissionsPlugin extends Plugin {

    private Timer monitorTimer;
    private Map<String, String> previousStates;
    private final PermissionDiffEngine diffEngine = new PermissionDiffEngine();

    @PluginMethod
    public void checkPermissions(PluginCall call) {
        call.resolve(buildPermissionsResponse());
    }

    @PluginMethod
    public void startMonitoring(PluginCall call) {
        stopTimer();
        // Seed the baseline so the first real change is detected (and we don't emit
        // a spurious event on start).
        previousStates = currentStates();
        monitorTimer = new Timer();
        monitorTimer.schedule(
            new TimerTask() {
                @Override
                public void run() {
                    emitIfChanged();
                }
            },
            10000,
            10000
        );
        call.resolve();
    }

    private void emitIfChanged() {
        Map<String, String> current = currentStates();
        List<PermissionDiffEngine.Change> changes = diffEngine.diff(previousStates, current);
        previousStates = current;
        if (changes.isEmpty()) return;

        JSObject ret = buildPermissionsResponse();
        ret.put("timestamp", System.currentTimeMillis());
        JSArray changesArr = new JSArray();
        for (PermissionDiffEngine.Change change : changes) {
            JSObject obj = new JSObject();
            obj.put("permission", change.permission);
            obj.put("from", change.from);
            obj.put("to", change.to);
            changesArr.put(obj);
        }
        ret.put("changes", changesArr);
        notifyListeners("permissionChange", ret);
    }

    @PluginMethod
    public void stopMonitoring(PluginCall call) {
        stopTimer();
        call.resolve();
    }

    private JSObject buildPermissionsResponse() {
        JSObject result = new JSObject();
        for (Map.Entry<String, String> entry : currentStates().entrySet()) {
            result.put(entry.getKey(), entry.getValue());
        }
        return result;
    }

    private Map<String, String> currentStates() {
        Map<String, PermissionState> states = getPermissionStates();
        Map<String, String> result = new HashMap<>();

        result.put("geolocation", stateToString(states.get("geolocation")));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            result.put("notifications", stateToString(states.get("notifications")));
        } else {
            // Pre-Android 13: notifications don't require runtime permission
            result.put("notifications", "granted");
        }

        result.put("notificationsPolicy", stateToString(states.get("notificationsPolicy")));
        return result;
    }

    private String stateToString(PermissionState state) {
        if (state == null) return "prompt";
        switch (state) {
            case GRANTED:
                return "granted";
            case DENIED:
                return "denied";
            default:
                return "prompt";
        }
    }

    private void stopTimer() {
        if (monitorTimer != null) {
            monitorTimer.cancel();
            monitorTimer = null;
        }
        previousStates = null;
    }

    @Override
    protected void handleOnDestroy() {
        stopTimer();
    }
}

File: ios/Plugin/DevicePermissionsPlugin.swift
import Foundation
import Capacitor

@objc(DevicePermissionsPlugin)
public class DevicePermissionsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DevicePermissionsPlugin"
    public let jsName = "DevicePermissions"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMonitoring", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopMonitoring", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = DevicePermissionsImpl()
    private let permissionKeys = ["geolocation", "notifications", "notificationsPolicy"]
    private var lastStatus: [String: String]?

    @objc public override func checkPermissions(_ call: CAPPluginCall) {
        implementation.checkPermissions { permissions in
            call.resolve(permissions)
        }
    }

    @objc func startMonitoring(_ call: CAPPluginCall) {
        // Seed the baseline so the first real change is detected (and we don't emit
        // a spurious event on start).
        implementation.checkPermissions { [weak self] initial in
            self?.lastStatus = initial
        }
        implementation.startMonitoring { [weak self] in
            guard let self = self else { return }
            self.implementation.checkPermissions { permissions in
                self.emitIfChanged(permissions)
            }
        }
        call.resolve()
    }

    private func emitIfChanged(_ current: [String: String]) {
        let previous = lastStatus
        lastStatus = current

        guard let previous = previous else { return }

        var changes: [[String: String]] = []
        for key in permissionKeys {
            let from = previous[key]
            let to = current[key]
            if let from = from, let to = to, from != to {
                changes.append(["permission": key, "from": from, "to": to])
            }
        }
        if changes.isEmpty { return }

        var data: [String: Any] = current
        data["timestamp"] = Date().timeIntervalSince1970 * 1000
        data["changes"] = changes
        notifyListeners("permissionChange", data: data)
    }

    @objc func stopMonitoring(_ call: CAPPluginCall) {
        implementation.stopMonitoring()
        lastStatus = nil
        call.resolve()
    }

    deinit {
        implementation.stopMonitoring()
    }
}
